atlasdb-timelock-server/src/main/java/com/palantir/atlasdb/timelock/paxos/PaxosServerImplementation.java

```diff
@@ -0,0 +1,233 @@
+import com.palantir.timestamp.TimestampService;
+
+import io.dropwizard.setup.Environment;
+
+public class PaxosServerImplementation implements ServerImplementation {
+    public static final String LEADER_NAMESPACE = "__leader";
+
+    private final PaxosConfiguration paxosConfiguration;
+    private final Environment environment;
+
+    private Set<String> remoteServers;
+    private Optional<SSLSocketFactory> optionalSecurity = Optional.absent();
+
+    private LeaderElectionService leaderElectionService;
+    private PaxosResource paxosResource;
+
+    public PaxosServerImplementation(PaxosConfiguration configuration, Environment environment) {
+        this.paxosConfiguration = configuration;
+        this.environment = environment;
+    }
+
+    @Override
+    public void onStartup(TimeLockServerConfiguration configuration) {
+        registerPaxosResource();
+
+        if (paxosConfiguration.sslConfiguration().isPresent()) {
+            optionalSecurity = constructOptionalSslSocketFactory(paxosConfiguration);
+        }
+
+        registerLeaderElectionService(configuration);
+        environment.jersey().register(new NotCurrentLeaderExceptionMapper());
+    }
+
+    private void registerPaxosResource() {
+        paxosResource = PaxosResource.create(paxosConfiguration.paxosDataDir().toString());
+        paxosResource.addClient(LEADER_NAMESPACE);
+        environment.jersey().register(paxosResource);
+    }
+
+    private void registerLeaderElectionService(TimeLockServerConfiguration configuration) {
+        remoteServers = getRemotePaths(configuration);
+        List<PaxosLearner> learners = getNamespacedProxies(
+                remoteServers,
+                LEADER_NAMESPACE,
+                optionalSecurity,
+                PaxosLearner.class);
+        learners.add(paxosResource.getPaxosLearner(LEADER_NAMESPACE));
+
+        List<PaxosAcceptor> acceptors = getNamespacedProxies(
+                remoteServers,
+                LEADER_NAMESPACE,
+                optionalSecurity,
+                PaxosAcceptor.class);
+        acceptors.add(paxosResource.getPaxosAcceptor(LEADER_NAMESPACE));
+
+        Map<PingableLeader, HostAndPort> otherLeaders = Leaders.generatePingables(
+                remoteServers,
+                optionalSecurity);
+
+        ExecutorService executor = Executors.newCachedThreadPool(new ThreadFactoryBuilder()
+                .setNameFormat("atlas-leaders-%d")
+                .setDaemon(true)
+                .build());
+
+        PaxosProposer proposer = PaxosProposerImpl.newProposer(
+                paxosResource.getPaxosLearner(LEADER_NAMESPACE),
+                ImmutableList.copyOf(acceptors),
+                ImmutableList.copyOf(learners),
+                getQuorumSize(configuration),
+                executor);
+
+        // Build and store a gatekeeping leader election service
+        leaderElectionService = new PaxosLeaderElectionService(
```

Comment:
This is a long list of constructor arguments to have to keep in the right order. Consider adding a builder.